### PR TITLE
Start migrating I/O writers to pylibcudf (starting with JSON)

### PR DESCRIPTION
**Pull request by lithomas1**
## Description

Switches the JSON writer to use pylibcudf.
xref #15162 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/_lib/json.pyx
Comment (mroeschke) [inline python/cudf/cudf/_lib/json.pyx:177]:
```suggestion
                    c.to_pylibcudf(mode="read") for c in table._columns
```
Minor nit

Comment (lithomas1) [reply python/cudf/cudf/_lib/json.pyx]: Thanks, updated.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (lithomas1) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:154]: This part is mostly copied and adapted from ``make_sink_info`` in cudf._lib.io.utils

File: python/cudf/cudf/pylibcudf_tests/test_json.py
Comment (lithomas1) [inline python/cudf/cudf/pylibcudf_tests/test_json.py:45]: I ended up not testing for round-tripping, since that doesn't always work for JSON.

e.g. when we have floats like 5.0, they get written by cudf (and pandas) as 5, and so get read back as ints.
(at least for the orient="records" mode)

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (mroeschke) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:217]: Is this branch supposed to be exactly the same as the `io.StringIO` case? If so would there be a way to combine this?

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (mroeschke) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:161]:
```suggestion
        if isinstance(buf, io.TextIO):
```
?

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (lithomas1) [reply python/cudf/cudf/_lib/pylibcudf/io/types.pyx]: Oh, I copied this verbatim, but this should be combinable.

I've updated it in the last commit.

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (wence-) [inline python/cudf/cudf/pylibcudf_tests/conftest.py:114]: *TODO (perhaps)*: This looks like a wonderful opportunity to use hypothesis to generate the data. i.e. we write rules for production of the base cases and combinators.

That would also allow generation of tests for `list(struct(T))` and `struct(list(T))` (for example).

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (wence-) [inline python/cudf/cudf/pylibcudf_tests/conftest.py:141]: Wah?! This is "once-only" fixture, no, so it should be cleaned up automatically. What is this for?

File: python/cudf/cudf/pylibcudf_tests/test_json.py
Comment (wence-) [inline python/cudf/cudf/pylibcudf_tests/test_json.py:31]: *question*: Is this a mismatch between cudf and pandas? If yes, we need to report a bug and fix it on the cudf side.

File: python/cudf/cudf/pylibcudf_tests/common/utils.py
Comment (wence-) [inline python/cudf/cudf/pylibcudf_tests/common/utils.py:141]: *issue*: Please add documentation.

File: python/cudf/cudf/pylibcudf_tests/test_json.py
Comment (wence-) [inline python/cudf/cudf/pylibcudf_tests/test_json.py:20]: Do we need this condition because if you pass `rows_per_chunk > num_rows` libcudf spits an error?

Comment (wence-) [reply python/cudf/cudf/pylibcudf_tests/test_json.py:20]: Or are you trying to test with default `rows_per_chunk=None`, if the latter, please add `None` as a parametrization.

**Review by wence- — CHANGES_REQUESTED** [review]
Thanks, I think there a few organisational tidy-ups we can do in the tests.

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (lithomas1) [reply python/cudf/cudf/pylibcudf_tests/conftest.py:51]: That's a good point. 

Will update my draft PR to include this.

I was following the convention set by the strings tests, which are also in the top level tests folder, so we should probably move those as well.

Comment (wence-) [reply python/cudf/cudf/pylibcudf_tests/conftest.py:51]: Yes, let me open a PR for that separately.

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (lithomas1) [reply python/cudf/cudf/pylibcudf_tests/conftest.py:141]: I couldn't figure out how to reset the StringIO buffers 😅 .
(otherwise we have the writes from the previous calls to write_json in it)

I tried setting the scope to "function", but that didn't seem to do anything.

File: python/cudf/cudf/pylibcudf_tests/test_json.py
Comment (lithomas1) [reply python/cudf/cudf/pylibcudf_tests/test_json.py:31]: No, orient="records" is just lossy.

e.g. 

```
>>> pd.DataFrame(columns=["a", "b"]).to_json(orient="records")
'[]'
```

File: python/cudf/cudf/pylibcudf_tests/test_json.py
Comment (lithomas1) [reply python/cudf/cudf/pylibcudf_tests/test_json.py:20]: > Do we need this condition because if you pass `rows_per_chunk > num_rows` libcudf spits an error?

Yes.

File: python/cudf/cudf/pylibcudf_tests/test_json.py
Comment (lithomas1) [reply python/cudf/cudf/pylibcudf_tests/test_json.py:31]: Actually, this is outdated from when I was trying to round-trip.

Let me try removing.

File: python/cudf/cudf/pylibcudf_tests/test_json.py
Comment (wence-) [reply python/cudf/cudf/pylibcudf_tests/test_json.py:20]: Perhaps that is a feature request, because that seems bad, I need to know the maximum number of rows in my file before I set rows_per_chunk.

File: python/cudf/cudf/pylibcudf_tests/common/utils.py
Comment (lithomas1) [reply python/cudf/cudf/pylibcudf_tests/common/utils.py:155]: Was planning on doing in a follow up, but this one seems easy, will be up in next commit.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (lithomas1) [reply python/cudf/cudf/_lib/pylibcudf/io/types.pyx:186]: Hm, maybe a good idea for a followup then, given that this is just the docstring?
(I'd love to be able to type this as List[IOSource] but we haven't started any typing stuff for pylibcudf)

I could also give up on the typing, and just write out the type as "homogenous list of str, bytesio, etc."

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (wence-) [reply python/cudf/cudf/pylibcudf_tests/conftest.py:141]: It should be the case that pytest conses a new object every time this fixture is used, so it seems something is wrong :(

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (vyasr) [reply python/cudf/cudf/pylibcudf_tests/conftest.py:116]: I could have sworn we had an issue about using hypothesis somewhere, but I can't find it. Maybe I never got around to documenting. Yes, I think using it would be great.

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (wence-) [reply python/cudf/cudf/pylibcudf_tests/conftest.py:143]: Can you distill this to a smaller test case, because this sounds absolutely like a pytest bug.

**Review by wence- — APPROVED** [review]
I don't have any more blocking comments I think. Thanks. I will let @vyasr decide if there are any more from their point of view.

My worry about needing to reset the StringIO bufs in the test fixture still applies. I absolutely don't think that should be necessary.

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (wence-) [reply python/cudf/cudf/pylibcudf_tests/conftest.py:143]: Ah ok, yes, I was blind to this. This is just standard python, because pytest only sees the actual objects already.

The right way to handle this is something like:

```
@pytest.fixture(params=["string", "pathlib.Path", "BytesIO", "StringIO"])
def source_or_sink(request, tmp_path):
    if request.param == "string":
        return f"{tmp_path}/{request.param}"
    elif request.param == "pathlib.Path":
        return pathlib.Path(tmp_path) / request.param
    elif request.param == "BytesIO":
        return io.BytesIO()
    elif request.param == "StringIO":
        return io.StringIO()
    assert False
```

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:58]: There are a lot of `move`s in this file. Are they really all necessary? For instance, this `move` seems like a no-op because this is just a string. The `metadata.schema_info` is a `vector[column_name_info]`, where `column_name_info` doesn't contain any `unique_ptr` or any other object that is movable (let alone not-copyable).

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:258]: There is a significant amount of duplicated logic in this function that I think could be simplified if instead of having 4 branches of the if/else you had two branches that each branched twice. The first branch would handle the io types, while the second branch would handle str/PathLike. Also, you may be able to do a single pass check of `all(isinstance(s, type(sinks[0]) for s in sinks`, again with a bit of special handling for the io types case though since `type(sinks[0])` won't be general enough.

Just a suggestion though, simplify as much as you can but it's not blocking.

File: python/cudf/cudf/pylibcudf_tests/common/utils.py
Comment (vyasr) [inline python/cudf/cudf/pylibcudf_tests/common/utils.py:149]: Maybe `nesting_level`? Just `nesting` could be returning a boolean, or it could be generating nesting, or any number of other things.

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (vyasr) [inline python/cudf/cudf/pylibcudf_tests/conftest.py:81]: Does this support a list of structs?

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (vyasr) [inline python/cudf/cudf/pylibcudf_tests/conftest.py:110]: Like above, does this support a struct of lists?

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (vyasr) [inline python/cudf/cudf/pylibcudf_tests/conftest.py:55]: Not a big deal in this (or most) contexts, but FYI [`{}` is always faster than `dict()`](https://madebyme.today/blog/python-dict-vs-curly-brackets/) so you may as well prefer it for initializing empty.

File: python/cudf/cudf/pylibcudf_tests/io/test_json.py
Comment (vyasr) [inline python/cudf/cudf/pylibcudf_tests/io/test_json.py:17]:
```suggestion
    kwargs = {}
```

File: python/cudf/cudf/pylibcudf_tests/test_copying.py
Comment (vyasr) [inline python/cudf/cudf/pylibcudf_tests/test_copying.py:28]: Wow this fixture got a lot more complex. I guess it's good to see that all the other existing tests pass with it. Are there other modules with fixtures that should be modified in this way? Not necessary in this PR, just wondering.

**Review by vyasr — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [reply python/cudf/cudf/_lib/pylibcudf/io/types.pyx:177]: You're right, looks like I was thinking of other memoryview functions and this one is fine. All good here.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [reply python/cudf/cudf/_lib/pylibcudf/io/types.pyx]: You generally only need to move vectors if those vectors contain something that requires a move (like a `unique_ptr`). There are cases where things don't require a move but benefit from the move from a perf perspective, but I don't think we have many of those here.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [reply python/cudf/cudf/_lib/pylibcudf/io/types.pyx:166]: It's a bit strange to define a `cppclass` in Cython. It would probably be better to switch it to a `cdef class` to avoid any oddness in semantics, but it's not a big deal.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:210]: Minor nit: since you know the type of sinks[0], if you do the conditional branching first and then loop you can check the type only once instead of once per element of sinks.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:208]: Do we have to deal with subclasses at all? Using `type` directly won't work in that case and you'd need to do a cascade of instance checks to test.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:214]: Another reason to do the for loop inside the conditional, you're reserving once per element which really isn't necessary.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:238]: You should be able to error earlier based on `initial_sink_cls`

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [reply python/cudf/cudf/_lib/pylibcudf/io/types.pyx:245]: Definitely better. I think it's worth one more small pass then we'll be good.

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (vyasr) [reply python/cudf/cudf/pylibcudf_tests/conftest.py:97]: Created an issue: https://github.com/rapidsai/cudf/issues/16129

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (vyasr) [reply python/cudf/cudf/pylibcudf_tests/conftest.py:51]: Did this PR get opened? If not, can we either make it, or at least open an issue so that we don't forget?

**Review by vyasr — CHANGES_REQUESTED** [review]
Couple of small things left, but enough that I think it's worth fixing before we merge

File: python/cudf/cudf/pylibcudf_tests/conftest.py
Comment (vyasr) [reply python/cudf/cudf/pylibcudf_tests/conftest.py:63]: Let's leave it as is for now.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:214]: You have to do this before the `initial_sink_cls` is computed and the check of types is done because you could have different PathLike representations at the same time and you need to parse them all to strings first.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:216]: Once you've preprocessed the pathlike objects, you can replace this `isinstance` check and the one below with just doing `if initial_sink_cls (is|in) ...` which will be a bit faster since you've already computed the type.

**Review by vyasr — APPROVED** [review]
Couple of last suggestions, but I'm approving now so feel free to merge once you've addressed as you see fit.

File: python/cudf/cudf/_lib/pylibcudf/io/types.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/pylibcudf/io/types.pyx:231]:
```suggestion
        elif initial_sink_cls is str:
```

**Pull request merged**
